### PR TITLE
feat: serialize on the stack

### PR DESCRIPTION
**Pull request by PastaPastaPasta**
This builds on top of #74 to enhance library performance when integrated into other stuff

### CODE REVIEW

File: src/elements.cpp [hunk 1]
```diff
@@ -182,6 +190,29 @@ std::vector<uint8_t> G1Element::Serialize(const bool fLegacy) const {
     return std::vector<uint8_t>(buffer + 1, buffer + 1 + G1Element::SIZE);
 }
 
+std::array<uint8_t, G1Element::SIZE> G1Element::SerializeToArray(const bool fLegacy) const {
+    uint8_t buffer[G1Element::SIZE + 1];
+    g1_write_bin(buffer, G1Element::SIZE + 1, p, 1);
+
+    if (buffer[0] == 0x00) {  // infinity
```
[the remark this answers is not in the document]
Comment (knst) [reply src/elements.cpp:197]: I think this one will work fine:

```
std::vector<uint8_t> G1Element::Serialize(const bool fLegacy) const {
  const auto arr = G1Element::SerializeToArray(fLegacy);
  return std::vector<uint8_t>{arr.begin(), arr.end()};
}
```

**Review by knst — CHANGES_REQUESTED** [review]

File: src/elements.cpp [hunk 2]
```diff
@@ -534,4 +611,11 @@ std::vector<uint8_t> GTElement::Serialize() const
     return data;
 }
 
+std::array<uint8_t, GTElement::SIZE> GTElement::SerializeToArray() const
+{
+    std::array<uint8_t, GTElement::SIZE> data{};
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/elements.cpp:602]: better to initialize data, then let it potentially be uninitialized

File: src/elements.cpp [hunk 3]
```diff
@@ -386,11 +394,18 @@ G2Element G2Element::Negate() const
 GTElement G2Element::Pair(const G1Element& a) const { return a & (*this); }
 
 std::vector<uint8_t> G2Element::Serialize(const bool fLegacy) const {
+  const auto arr = G2Element::SerializeToArray(fLegacy);
+  return std::vector<uint8_t>{arr.begin(), arr.end()};
```
Comment (UdjinM6) [inline src/elements.cpp:398]: nit: whitespaces
```suggestion
    const auto arr = G2Element::SerializeToArray(fLegacy);
    return std::vector<uint8_t>{arr.begin(), arr.end()};
```

File: src/elements.cpp [hunk 4]
```diff
@@ -171,11 +171,16 @@ uint32_t G1Element::GetFingerprint(const bool fLegacy) const
 }
 
 std::vector<uint8_t> G1Element::Serialize(const bool fLegacy) const {
+  const auto arr = G1Element::SerializeToArray(fLegacy);
+  return std::vector<uint8_t>{arr.begin(), arr.end()};
```
Comment (UdjinM6) [inline src/elements.cpp:175]: nit: whitespaces
```suggestion
    const auto arr = G1Element::SerializeToArray(fLegacy);
    return std::vector<uint8_t>{arr.begin(), arr.end()};
```

File: src/elements.cpp [hunk 5]
```diff
@@ -171,11 +171,16 @@ uint32_t G1Element::GetFingerprint(const bool fLegacy) const
...
+}
+
+std::array<uint8_t, G1Element::SIZE> G1Element::SerializeToArray(const bool fLegacy) const {
     uint8_t buffer[G1Element::SIZE + 1];
     g1_write_bin(buffer, G1Element::SIZE + 1, p, 1);
 
     if (buffer[0] == 0x00) {  // infinity
-        std::vector<uint8_t> result(G1Element::SIZE, 0);
+        std::array<uint8_t, G1Element::SIZE> result{};
```
Comment (UdjinM6) [inline src/elements.cpp:183]: maybe move it out of `if` like i's done in `G2Element::SerializeToArray()`?

**Comment by PastaPastaPasta** [comment]
All done

**Review by UdjinM6 — APPROVED** [review]
utACK 30aa085b2b8a169c157d3dfcba2caf604e8d3e54

File: src/privatekey.cpp [hunk 6]
```diff
@@ -284,6 +284,13 @@ std::vector<uint8_t> PrivateKey::Serialize(const bool fLegacy) const
     return data;
 }
 
+std::array<uint8_t, PrivateKey::PRIVATE_KEY_SIZE> PrivateKey::SerializeToArray(bool fLegacy) const
+{
+    std::array<uint8_t, PRIVATE_KEY_SIZE> data{};
+    Serialize(data.data());
```
Comment (knst) [inline src/privatekey.cpp:290]: is it forgotten to use `fLegacy` here?

Comment (PastaPastaPasta) [reply src/privatekey.cpp:290]: Not sure, but I'm guessing no? both PrivateKey::Serialize(uint8_t *buffer) and PrivateKey::Serialize(const bool fLegacy) don't use it. Maybe. private keys are not affected by legacy / non legacy

File: src/elements.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (knst) [reply src/elements.cpp:569]: yes, indeed, it set array to zeroes in this specific syntax, C++ magic.
It has a constructor from initialized list, and this line is an equivalent of `<c-array> = {}`

**Review by knst — APPROVED** [review]
utACK 30aa085b2b8a169c157d3dfcba2caf604e8d3e54

**Review by kwvg — APPROVED** [review]
ACK 30aa085b2b8a169c157d3dfcba2caf604e8d3e54

---

Tested on `develop` ([`1e553102`](https://github.com/dashpay/dash/tree/1e55310232dc000570438dc5fe981326b695fd79)) by updating subtree, unit and functional tests passed (see branch [`test_bls_75`](https://github.com/kwvg/dash/tree/cd115c29896fcee9f04b893d5c469f824c33afcd)).

File: src/privatekey.cpp (hunk 6 above)
Comment (UdjinM6) [reply src/privatekey.cpp:290]: Yes, it's a dummy one here which is needed because our `CBLSWrapper` expects it to be that way for every `ImplType`.

**Review by UdjinM6 — APPROVED** [review]
utACK 30aa085b2b8a169c157d3dfcba2caf604e8d3e54

**Pull request merged**